pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,251 @@
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		// it looks like we have found a member that isn't backed by a machine nor a node
+		// issue a live request to the etcd cluster and remove it from the cluster
+		liveMemberList, err := c.etcdClient.MemberList(ctx)
+		if err != nil {
+			return err
+		}
+		for _, liveMember := range liveMemberList {
+			liveMemberIP, err := memberToNodeInternalIP(liveMember)
+			if err != nil {
+				return err
+			}
+			if liveMemberIP != potentialMemberToRemoveIP {
+				continue
+			}
+			memberLocator = fmt.Sprintf("[ url: %v, name: %v, id: %v ]", liveMemberIP, liveMember.Name, liveMember.ID)
+			if len(liveMember.Name) == 0 {
+				// TODO: what if the member hasn't been started and the name field is empty ?!
+				klog.Errorf("unable to remove member: %v because it doesn't have a name (has the member been started?)", memberLocator)
+				break
+			}
+			err = c.etcdClient.MemberRemove(ctx, liveMember.Name)
+			if err != nil {
+				klog.Errorf("failed to remove member: %v, err: %v", memberLocator, err)
+			} else {
+				klog.V(2).Infof("successfully removed member: %v from the cluster", memberLocator)
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ClusterMemberRemovalController) getMachineForMember(memberInternalIP string) (*machinev1beta1.Machine, error) {
+	node, err := c.getNodeForMember(memberInternalIP)
+	if err != nil {
+		return nil, err
+	}
+
+	// node is gone, make sure we don't have a machine with
+	// the same internal IP as the node would have had
+	if node == nil {
+		nodeInternalIP := memberInternalIP
+		machine, err := c.findMachineByNodeInternalIP(nodeInternalIP)
+		if err != nil {
+			return nil, err
+		}
+		return machine, nil
+	}
+
+	machines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return nil, err
+	}
+	for _, machine := range machines {
+		if machine.Status.NodeRef != nil && machine.Status.NodeRef.Name == node.Name {
+			return machine, nil
+		}
+	}
+	return nil, nil
+}
+
+func (c *ClusterMemberRemovalController) getNodeForMember(memberInternalIP string) (*corev1.Node, error) {
+	masterNodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	network, err := c.networkLister.Get("cluster")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster network: %w", err)
+	}
+
+	for _, masterNode := range masterNodes {
+		internalNodeIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, masterNode)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get internal IP for node: %w", err)
+		}
+		if memberInternalIP == internalNodeIP {
+			return masterNode, nil
+		}
+	}
+	return nil, nil
+}
+
+// findMachineByNodeInternalIP finds machine that matches the given nodeInternalIP
+// is safe because the MCO:
```

Comment:
MCO (Machine Config Operator) is not involved here is it? Do you mean MAO? (Machine API Operator)

Reply:
yes, thx